### PR TITLE
Dynamic pods: support for loading new datasets

### PR DESCRIPTION
**Pull request by sgrebnov**
1. Extracted dataset loading logic to re-usable function so it cab used by pods watcher
2. Pods watcher is now async  so it can work with datasets loading and other async functionality
3. Datafusion is shared via RWLock

### CODE REVIEW

File: bin/spiced/src/lib.rs [hunk 1]
```diff
@@ -99,65 +77,9 @@ pub async fn run(args: Args) -> Result<()> {
...
-                        UnableToAttachDataConnectorSnafu {
-                            data_connector: source,
-                        },
-                    )?;
-                }
-            }
-        }
-
-        tracing::info!("Loaded dataset: {}", ds.name);
+        Runtime::load_dataset(ds, &auth, &mut df)
```
Comment (mitchdevenport) [inline bin/spiced/src/lib.rs:77]: Looks like you have a conflict to resolve, I think this code was changed recently.

Comment (sgrebnov) [reply bin/spiced/src/lib.rs:77]: @mitchdevenport - merged the latest from `rust`, thanks!

File: crates/runtime/src/lib.rs [hunk 2]
```diff
@@ -66,7 +105,75 @@ impl Runtime {
         }
     }
 
-    pub async fn start_servers(&self) -> Result<()> {
+    pub async fn load_dataset(
```
Comment (mitchdevenport) [inline crates/runtime/src/lib.rs:111]: nit: should be called `load_datasets`

Comment (mitchdevenport) [reply crates/runtime/src/lib.rs:111]: Oh wait, I misread

File: crates/runtime/src/lib.rs [hunk 3]
```diff
@@ -76,35 +183,41 @@ impl Runtime {
...
+        while let Some(new_app) = rx.recv().await {
+            tracing::debug!("Updated pods information: {:?}", new_app);
+            tracing::debug!("Previous pods information: {:?}", current_app);
 
-                    current_app = Arc::new(new_app);
-                }
-            }
-        };
+            // TODO: This is a placeholder to demo/test that datasets could be loaded here
+            // To be replaced with actual implementation using shared ds/datafusion instance
```
Comment (mitchdevenport) [inline crates/runtime/src/lib.rs]: Why can't we add this now?

Comment (sgrebnov) [reply crates/runtime/src/lib.rs]: @mitchdevenport  - added

File: crates/runtime/src/lib.rs [hunk 4]
```diff
@@ -36,14 +37,53 @@ pub enum Error {
 
     #[snafu(display("Unable to start OpenTelemetry server"))]
     UnableToStartOpenTelemetryServer { source: opentelemetry::Error },
+
```
Comment (mitchdevenport) [inline crates/runtime/src/lib.rs:40]: All these errors need to explicitly print the underlying source. I just merged a PR that went through and fixed this for every error in the codebase so you just need to make sure you pull in those changes properly.

**Review by mitchdevenport — APPROVED** [review]

File: crates/runtime/src/lib.rs (hunk 4 above)
Comment (sgrebnov) [reply crates/runtime/src/lib.rs:40]: Thank you - double checked all errors are propogated

**Pull request merged**